chain/chain/src/resharding.rs

```diff
@@ -0,0 +1,182 @@
+    pub state_root: StateRoot,
+    pub next_epoch_shard_layout: ShardLayout,
+    pub state_split_status: Arc<StateSplitApplyingStatus>,
+}
+
+#[derive(actix::Message)]
+#[rtype(result = "()")]
+pub struct StateSplitResponse {
+    pub sync_hash: CryptoHash,
+    pub shard_id: ShardId,
+    pub new_state_roots: Result<HashMap<ShardUId, StateRoot>, Error>,
+}
+
+fn apply_delayed_receipts<'a>(
+    tries: &ShardTries,
+    orig_shard_uid: ShardUId,
+    orig_state_root: StateRoot,
+    state_roots: HashMap<ShardUId, StateRoot>,
+    account_id_to_shard_id: &(dyn Fn(&AccountId) -> ShardUId + 'a),
+) -> Result<HashMap<ShardUId, StateRoot>, Error> {
+    let orig_trie_update = tries.new_trie_update_view(orig_shard_uid, orig_state_root);
+
+    let mut start_index = None;
+    let mut new_state_roots = state_roots;
+    while let Some((next_index, receipts)) =
+        get_delayed_receipts(&orig_trie_update, start_index, STATE_PART_MEMORY_LIMIT)?
+    {
+        let (store_update, updated_state_roots) = tries.apply_delayed_receipts_to_split_states(
+            &new_state_roots,
+            &receipts,
+            account_id_to_shard_id,
+        )?;
+        new_state_roots = updated_state_roots;
+        start_index = Some(next_index);
+        store_update.commit()?;
+    }
+
+    Ok(new_state_roots)
+}
+
+impl Chain {
+    pub fn build_state_for_split_shards_preprocessing(
+        &self,
+        sync_hash: &CryptoHash,
+        shard_id: ShardId,
+        state_split_scheduler: &dyn Fn(StateSplitRequest),
+        state_split_status: Arc<StateSplitApplyingStatus>,
+    ) -> Result<(), Error> {
+        let (epoch_id, next_epoch_id) = {
+            let block_header = self.get_block_header(sync_hash)?;
+            (block_header.epoch_id().clone(), block_header.next_epoch_id().clone())
+        };
+        let shard_layout = self.epoch_manager.get_shard_layout(&epoch_id)?;
+        let next_epoch_shard_layout = self.epoch_manager.get_shard_layout(&next_epoch_id)?;
+        let shard_uid = ShardUId::from_shard_id_and_layout(shard_id, &shard_layout);
+        let prev_hash = *self.get_block_header(sync_hash)?.prev_hash();
+        let state_root = *self.get_chunk_extra(&prev_hash, &shard_uid)?.state_root();
+        assert_ne!(shard_layout, next_epoch_shard_layout);
+
+        state_split_scheduler(StateSplitRequest {
+            runtime_adapter: self.runtime_adapter.clone(),
+            sync_hash: *sync_hash,
+            shard_id,
+            shard_uid,
+            state_root,
+            next_epoch_shard_layout,
+            state_split_status,
+        });
+
+        Ok(())
+    }
+
+    pub fn build_state_for_split_shards(
+        state_split_request: StateSplitRequest,
+    ) -> Result<HashMap<ShardUId, StateRoot>, Error> {
+        let StateSplitRequest {
+            runtime_adapter,
+            shard_uid,
+            state_root,
+            next_epoch_shard_layout,
+            state_split_status,
+            ..
```

Comment:
strange that we don't unpack all args 
ditto - not to be fixed here but would be nice to understand what's up

Reply:
The arguments are used in functions that call build_state_for_split_shards as part of response I think.